pkg/venafi/firefly/deviceFlow.go

```diff
@@ -0,0 +1,197 @@
+	}
+
+	fmt.Printf("please go to %v and enter the code %v . It will expire in %dm and %ds\n\n", verificationURL, devCred.UserCode, devCred.ExpiresIn/60, devCred.ExpiresIn%60)
+
+	//waiting for the user authorization
+	return c.waitForDeviceAuthorization(devCred, auth)
+}
+
+func (c *Connector) requestDeviceCode(auth *endpoint.Authentication) (*DeviceCred, error) {
+	data := url.Values{
+		"client_id": {auth.ClientId},
+	}
+
+	//There are IdPs like Okta and Auth0 which provides the support for default scopes, so it's possible for
+	//these that the scope is empty
+	if auth.Scope != "" {
+		data.Add("scope", auth.Scope)
+	}
+
+	//audience is only supported by Okta and Auth0
+	if auth.IdentityProvider.Audience != "" {
+		data.Add("audience", auth.IdentityProvider.Audience)
+	}
+
+	statusCode, status, body, err := c.request("POST", urlResource(auth.IdentityProvider.DeviceURL), data)
+
+	if err != nil {
+		return nil, err
+	}
+	//parsing the response
+	return parseDeviceCodeRequestResult(statusCode, status, body)
+}
+
+func parseDeviceCodeRequestResult(httpStatusCode int, httpStatus string, body []byte) (*DeviceCred, error) {
+	switch httpStatusCode {
+	case http.StatusOK:
+		return parseDeviceCodeRequestData(body)
+	default:
+		respError, err := NewResponseError(body)
+		if err != nil {
+			return nil, err
+		}
+
+		return nil, fmt.Errorf("unexpected status code requesting Device Code. Status: %s error: %w", httpStatus, respError)
+	}
+}
+
+func parseDeviceCodeRequestData(b []byte) (*DeviceCred, error) {
+	var data DeviceCred
+	err := json.Unmarshal(b, &data)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", verror.ServerError, err)
+	}
+
+	return &data, nil
+}
+
+func (c *Connector) waitForDeviceAuthorization(devCred *DeviceCred, auth *endpoint.Authentication) (*oauth2.Token, error) {
+
+	data := url.Values{"client_id": {auth.ClientId},
+		"device_code": {devCred.DeviceCode},
+		"grant_type":  {"urn:ietf:params:oauth:grant-type:device_code"},
+	}
+
+	// Google requires the client-secret also to request the accessToken
+	if auth.ClientSecret != "" {
+		data.Add("client_secret", auth.ClientSecret)
+	}
+
+	//polling the authorization
+	for {
+		//requesting the authorization
+		statusCode, _, body, err := c.request("POST", urlResource(auth.IdentityProvider.TokenURL), data)
+		if err != nil {
+			return nil, err
+		}
+
+		//parsing the response
+		token, err := parseWaitingDeviceAuthorizationRequestResult(statusCode, body)
+
+		if err != nil {
+			switch err.Error() {
+			case "authorization_pending":
+				time.Sleep(time.Duration(devCred.Interval) * time.Second)
+			case "slow_down":
+				devCred.Interval += 5
+				time.Sleep(time.Duration(devCred.Interval) * time.Second)
+			case "access_denied":
+				return nil, fmt.Errorf("the access from device was denied by the user")
+			case "expired_token":
+				return nil, fmt.Errorf("the device code expired")
+			default:
+				return nil, fmt.Errorf("the authorization failed. %w", err)
+			}
+		} else {
```

Comment:
invert this if/else to remove nesting. If err == nil return with token. Then no else is necessary and you can follow the next steps to check the error and the switch.

Reply:
done